File: ds/js/set.js
function MySet() {
  const collection = [];

  this.has = function(element) {
    return collection.indexOf(element) !== -1;
  }

  this.add = function(element) {
    if (this.has(element)) {
      return false;
    }

    collection.push(element);
    return true;
  }

  this.remove = function(element) {
    if (!this.has(element)) {
      return false;
    }

    const position = collection.indexOf(element);
    collection.splice(position, 1);

    return true;
  }

  this.values = function () {
    return collection;
  }

  this.size = function() {
    return collection.length;
  }

  this.union = function(otherSet) {
    const unionSet = new MySet();
    const currentSetValues = this.values();
    const otherSetValues = otherSet.values();

    currentSetValues.forEach((el) => {
      unionSet.add(el);
    });

    otherSetValues.forEach((el) => {
      unionSet.add(el);
    });

    return unionSet;
  }

  this.intersection = function (otherSet) {
    const intersectionSet = new MySet();
    const otherSetValues = otherSet.values();

    otherSetValues.forEach((el) => {
      if (this.has(el)) {
        intersectionSet.add(el);
      }
    });

    return intersectionSet;
  }

  this.difference = function(otherSet) {
    const differenceSet = new MySet();
    const currentSetValues = this.values();

    currentSetValues.forEach((el) => {
      if (!otherSet.has(el)) {
        differenceSet.add(el);
      }
    });

    return differenceSet;
  }
}

const set = new MySet();
set.add(10);
set.add(15);
set.add(25);

const anotherSet = new MySet();
anotherSet.add(10);
anotherSet.add(30);

console.log(set.values());
console.log(set.union(anotherSet).values());
console.log(set.intersection(anotherSet).values());
console.log(set.difference(anotherSet).values());

File: ds/golang/set/set.go
package main

// Set structure
// Time complexity depends on DS of storage. Current implementation uses simple array
// For optimization we can use `balanced tree`
type Set struct {
	data   []int
	length int
}

// Add appends new element to the set
// Time complexity: O(N)
func (s *Set) Add(x int) bool {
	if s.Has(x) {
		return false
	}

	s.data = append(s.data, x)
	s.length++

	return true
}

// Has checks if Set contains element
func (s *Set) Has(x int) bool {
	for _, v := range s.data {
		if v == x {
			return true
		}
	}

	return false
}

// Values returns Sets' data
func (s *Set) Values() []int {
	return s.data
}

// Union merges two Sets
// Time complexity: O(max(N, M)), where N, M are Sets size
func (s *Set) Union(anotherSet Set) Set {
	unionSet := Set{}

	for _, v := range s.Values() {
		unionSet.Add(v)
	}

	for _, v := range anotherSet.Values() {
		unionSet.Add(v)
	}

	return unionSet
}

// Intersection returns intersection between two Sets
// Time complexity: O(N*M)
func (s *Set) Intersection(anotherSet Set) Set {
	intersectionSet := Set{}
	for _, v := range s.Values() {
		if anotherSet.Has(v) {
			intersectionSet.Add(v)
		}
	}

	return intersectionSet
}

// Difference returns difference between two Sets
// Time complexity: O(N*M)
func (s *Set) Difference(anotherSet Set) Set {
	differenceSet := Set{}
	for _, v := range s.Values() {
		if !anotherSet.Has(v) {
			differenceSet.Add(v)
		}
	}

	return differenceSet
}
